# King County Home Price Predictors - An Analysis

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

from statsmodels.formula.api import ols
import statsmodels.api as sm
import scipy.stats as stats

import os
RPSale = pd.read_csv('../../data/raw/EXTR_RPSale.csv')
ResBldg = pd.read_csv('../../data/raw/EXTR_ResBldg.csv')
Parcel = pd.read_csv('../../data/raw/EXTR_Parcel.csv')

# Filtering the Data

### Filtering to The Last 3 Years

In [2]:
Date = RPSale['DocumentDate'].str.split('/', expand=True)
RPSale['MonthSold'] = Date[0].astype('float64')
RPSale['YrSold'] = Date[2].astype('float64')

#Drop original date column
RPSale.drop(columns=['DocumentDate'], axis=1, inplace=True)

In [3]:
#filtering data from 2018, 2019 and 2020 only
years = [2018, 2019, 2020]  
    
properties = RPSale.loc[RPSale['YrSold'].isin(years)]  
    
properties

ExciseTaxNbr   Major  Minor  SalePrice    RecordingNbr Volume Page  \
2            2999169  919715    200     192000  20190712001080               
8            3000673  894444    200     185000  20190722001395               
9            3027422  213043    120     560000  20191226000848               
10           3002257  940652    630     435000  20190730001339               
19           3018109  152504   9008    7600000  20191030001615               
...              ...     ...    ...        ...             ...    ...  ...   
351056       2997920  302300    320          0  20190705000539               
351057       3028691  277110   4239     955000  20200103000685               
351058       2934041  388580   3260    1375000  20180604000929               
351062       2935450  219331    270     850000  20180612000762               
351065       2971374  924600    650          0                               

       PlatNbr PlatType         PlatLot  ... SaleInstrument AFForestLand  \
2                                        ...              3            N   
8                                        ...              3            N   
9                                        ...              3            N   
10                                       ...              3            N   
19                                       ...              3            N   
...        ...      ...             ...  ...            ...          ...   
351056                                   ...             15            N   
351057                                   ...              3            N   
351058                                   ...              3            N   
351062                                   ...              3            N   
351065                                   ...              3            N   

       AFCurrentUseLand  AFNonProfitUse  AFHistoricProperty  SaleReason  \
2                     N               N                   N           1   
8                     N               N                   N           1   
9                     N               N                   N           1   
10                    N               N                   N           1   
19                    N               N                   N           1   
...                 ...             ...                 ...         ...   
351056                N               N                   N           1   
351057                N               N                   N           1   
351058                N               N                   N           1   
351062                N               N                   N           1   
351065                N               N                   N           1   

       PropertyClass SaleWarning MonthSold  YrSold  
2                  3                   7.0  2019.0  
8                  3                   6.0  2019.0  
9                  8                  12.0  2019.0  
10                 8                   7.0  2019.0  
19                 2                  10.0  2019.0  
...              ...         ...       ...     ...  
351056             8                   5.0  2019.0  
351057             8                  12.0  2019.0  
351058             8                   5.0  2018.0  
351062             8                   6.0  2018.0  
351065             8                   1.0  2019.0  

[143177 rows x 25 columns]

In [4]:
properties['YrSold'].value_counts()

2019.0    61351
2018.0    61170
2020.0    20656
Name: YrSold, dtype: int64

In [5]:
properties = properties.rename(columns=str.lower)

In [6]:
properties.head()

excisetaxnbr   major  minor  saleprice    recordingnbr volume page  \
2        2999169  919715    200     192000  20190712001080               
8        3000673  894444    200     185000  20190722001395               
9        3027422  213043    120     560000  20191226000848               
10       3002257  940652    630     435000  20190730001339               
19       3018109  152504   9008    7600000  20191030001615               

   platnbr plattype         platlot  ... saleinstrument afforestland  \
2                                    ...              3            N   
8                                    ...              3            N   
9                                    ...              3            N   
10                                   ...              3            N   
19                                   ...              3            N   

   afcurrentuseland  afnonprofituse  afhistoricproperty  salereason  \
2                 N               N                   N           1   
8                 N               N                   N           1   
9                 N               N                   N           1   
10                N               N                   N           1   
19                N               N                   N           1   

   propertyclass salewarning monthsold  yrsold  
2              3                   7.0  2019.0  
8              3                   6.0  2019.0  
9              8                  12.0  2019.0  
10             8                   7.0  2019.0  
19             2                  10.0  2019.0  

[5 rows x 25 columns]

### Padding the 'major' and 'minor' columns and merging them together:

In [7]:
properties['major'] = properties['major'].astype(str)
properties['minor'] = properties['minor'].astype(str)

def major_padded(row):
    if len(row) == 6:
        return row
    elif len(row) == 5:
        return '0'+row
    elif len(row) == 4:
        return '00'+row
    elif len(row) == 3:
        return '000'+row
    elif len(row) == 2:
        return '0000'+row
    elif len(row) == 1:
        return '00000'+row
    else:
        return '000000'

def minor_padded(row):
    if len(row) == 4:
        return row
    elif len(row) == 3: 
        return '0'+row
    elif len(row) == 2:
        return '00'+row
    elif len(row) == 1:
        return '000'+row
    else:
        return '0000'
    
properties['major'] = properties.major.apply(major_padded)
properties['minor'] = properties.minor.apply(minor_padded)
properties['major_minor'] = properties['major']+properties['minor']

In [8]:
#Dropping the old columns
to_drop = ['major','minor']
properties.drop(to_drop, axis=1, inplace=True)

In [9]:
#Moving the 'major_minor' column to the front
col_name = 'major_minor'
last_col = properties.pop(col_name)
properties.insert(1, col_name, last_col)

### Narrowing the property type
According to our LookUp file, the residental values are coded as follows:

**11** - Household, single family units

**12** - Multiple family residence (Residential, 2-4 units)

**14** - Residential condominiums

**18** - All other residential not elsewhere coded

In [10]:
properties = properties[properties['propertytype'].isin([11, 12, 14, 18])]
properties

excisetaxnbr major_minor  saleprice    recordingnbr volume page  \
9            3027422  2130430120     560000  20191226000848               
10           3002257  9406520630     435000  20190730001339               
22           3015516  7797900030          0  20191016000009               
24           2963345  7574800320          0  20181126001127               
27           3033993  9264902060     370000  20200214001331               
...              ...         ...        ...             ...    ...  ...   
351043       3045715  4127000898     619000  20200505001286               
351047       2949086  3597000405          0  20180824000032               
351056       2997920  3023000320          0  20190705000539               
351062       2935450  2193310270     850000  20180612000762               
351065       2971374  9246000650          0                               

       platnbr plattype         platlot platblock  ... saleinstrument  \
9                                                  ...              3   
10                                                 ...              3   
22                                                 ...              3   
24                                                 ...             15   
27                                                 ...              2   
...        ...      ...             ...       ...  ...            ...   
351043                                             ...              2   
351047                                             ...             15   
351056                                             ...             15   
351062                                             ...              3   
351065                                             ...              3   

       afforestland  afcurrentuseland  afnonprofituse  afhistoricproperty  \
9                 N                 N               N                   N   
10                N                 N               N                   N   
22                N                 N               N                   N   
24                N                 N               N                   N   
27                N                 N               N                   N   
...             ...               ...             ...                 ...   
351043            N                 N               N                   N   
351047            N                 N               N                   N   
351056            N                 N               N                   N   
351062            N                 N               N                   N   
351065            N                 N               N                   N   

       salereason propertyclass salewarning monthsold  yrsold  
9               1             8                  12.0  2019.0  
10              1             8                   7.0  2019.0  
22              1             8                  10.0  2019.0  
24              1             3                  11.0  2018.0  
27              1             8                   2.0  2020.0  
...           ...           ...         ...       ...     ...  
351043          1             8                   4.0  2020.0  
351047          1             8                   7.0  2018.0  
351056          1             8                   5.0  2019.0  
351062          1             8                   6.0  2018.0  
351065          1             8                   1.0  2019.0  

[68394 rows x 24 columns]

### Only using sale prices over 0:

In [11]:
properties = properties[properties['saleprice'] > 0]
properties

excisetaxnbr major_minor  saleprice    recordingnbr volume page  \
9            3027422  2130430120     560000  20191226000848               
10           3002257  9406520630     435000  20190730001339               
27           3033993  9264902060     370000  20200214001331               
42           3043271  5703500010     575000  20200417000249               
50           2937935  8731780100     185000  20180625000560               
...              ...         ...        ...             ...    ...  ...   
351038       3025068  6150200505     745000  20191212000602               
351040       2986698  0333100255    6500000  20190508000847               
351041       3038910  0267700310     455000  20200317001112               
351043       3045715  4127000898     619000  20200505001286               
351062       2935450  2193310270     850000  20180612000762               

       platnbr plattype         platlot platblock  ... saleinstrument  \
9                                                  ...              3   
10                                                 ...              3   
27                                                 ...              2   
42                                                 ...              2   
50                                                 ...              3   
...        ...      ...             ...       ...  ...            ...   
351038                                             ...              3   
351040                                             ...              3   
351041                                             ...              2   
351043                                             ...              2   
351062                                             ...              3   

       afforestland  afcurrentuseland  afnonprofituse  afhistoricproperty  \
9                 N                 N               N                   N   
10                N                 N               N                   N   
27                N                 N               N                   N   
42                N                 N               N                   N   
50                N                 N               N                   N   
...             ...               ...             ...                 ...   
351038            N                 N               N                   N   
351040            N                 N               N                   N   
351041            N                 N               N                   N   
351043            N                 N               N                   N   
351062            N                 N               N                   N   

       salereason propertyclass salewarning monthsold  yrsold  
9               1             8                  12.0  2019.0  
10              1             8                   7.0  2019.0  
27              1             8                   2.0  2020.0  
42              1             8                   4.0  2020.0  
50              1             3                   6.0  2018.0  
...           ...           ...         ...       ...     ...  
351038          1             8                  11.0  2019.0  
351040          1             8                   4.0  2019.0  
351041          1             3                   3.0  2020.0  
351043          1             8                   4.0  2020.0  
351062          1             8                   6.0  2018.0  

[47760 rows x 24 columns]

In [12]:
properties['saleprice'].describe()

count    4.776000e+04
mean     7.446527e+05
std      6.547413e+05
min      5.000000e+00
25%      4.250000e+05
50%      6.180000e+05
75%      8.680000e+05
max      5.300000e+07
Name: saleprice, dtype: float64

### Filtering out based on principal use:

In [13]:
properties['principaluse'].unique()

array([6, 2, 7, 0], dtype=int64)

According to our lookup, 

**6** - Residential

**2** - Condominium

**7** - Commercial

I will only focus on residential and condo properties.

In [14]:
properties = properties[properties['principaluse'].isin([6,2])]
properties

excisetaxnbr major_minor  saleprice    recordingnbr volume page  \
9            3027422  2130430120     560000  20191226000848               
10           3002257  9406520630     435000  20190730001339               
27           3033993  9264902060     370000  20200214001331               
42           3043271  5703500010     575000  20200417000249               
50           2937935  8731780100     185000  20180625000560               
...              ...         ...        ...             ...    ...  ...   
351038       3025068  6150200505     745000  20191212000602               
351040       2986698  0333100255    6500000  20190508000847               
351041       3038910  0267700310     455000  20200317001112               
351043       3045715  4127000898     619000  20200505001286               
351062       2935450  2193310270     850000  20180612000762               

       platnbr plattype         platlot platblock  ... saleinstrument  \
9                                                  ...              3   
10                                                 ...              3   
27                                                 ...              2   
42                                                 ...              2   
50                                                 ...              3   
...        ...      ...             ...       ...  ...            ...   
351038                                             ...              3   
351040                                             ...              3   
351041                                             ...              2   
351043                                             ...              2   
351062                                             ...              3   

       afforestland  afcurrentuseland  afnonprofituse  afhistoricproperty  \
9                 N                 N               N                   N   
10                N                 N               N                   N   
27                N                 N               N                   N   
42                N                 N               N                   N   
50                N                 N               N                   N   
...             ...               ...             ...                 ...   
351038            N                 N               N                   N   
351040            N                 N               N                   N   
351041            N                 N               N                   N   
351043            N                 N               N                   N   
351062            N                 N               N                   N   

       salereason propertyclass salewarning monthsold  yrsold  
9               1             8                  12.0  2019.0  
10              1             8                   7.0  2019.0  
27              1             8                   2.0  2020.0  
42              1             8                   4.0  2020.0  
50              1             3                   6.0  2018.0  
...           ...           ...         ...       ...     ...  
351038          1             8                  11.0  2019.0  
351040          1             8                   4.0  2019.0  
351041          1             3                   3.0  2020.0  
351043          1             8                   4.0  2020.0  
351062          1             8                   6.0  2018.0  

[47404 rows x 24 columns]

### Checking for 'propertyclass'

In [15]:
properties['propertyclass'].unique()

array([8, 3, 7, 9], dtype=int64)

The follwing values and their definitions are found in our lookup file:

**7** - Res-Land only (I will remove this to only display residential property)

I will keep:

**8** - Residential- Improved Property

**3** - Condominium

**9** - Res/Condo mobile

In [16]:
properties = properties[properties['propertyclass'].isin([8,3,9])]
properties.head()

excisetaxnbr major_minor  saleprice    recordingnbr volume page platnbr  \
9        3027422  2130430120     560000  20191226000848                       
10       3002257  9406520630     435000  20190730001339                       
27       3033993  9264902060     370000  20200214001331                       
42       3043271  5703500010     575000  20200417000249                       
50       2937935  8731780100     185000  20180625000560                       

   plattype         platlot platblock  ... saleinstrument afforestland  \
9                                      ...              3            N   
10                                     ...              3            N   
27                                     ...              2            N   
42                                     ...              2            N   
50                                     ...              3            N   

    afcurrentuseland  afnonprofituse  afhistoricproperty salereason  \
9                  N               N                   N          1   
10                 N               N                   N          1   
27                 N               N                   N          1   
42                 N               N                   N          1   
50                 N               N                   N          1   

   propertyclass salewarning monthsold  yrsold  
9              8                  12.0  2019.0  
10             8                   7.0  2019.0  
27             8                   2.0  2020.0  
42             8                   4.0  2020.0  
50             3                   6.0  2018.0  

[5 rows x 24 columns]

### Checking for duplicates based on major_minor numbers

In [17]:
len(properties) - len(properties.major_minor.unique())

1585

I will check to see if there are any trends in the duplicates, since it is possible a home could have been sold more than once during the set timeframe.

In [18]:
properties[properties['major_minor'].isin(['8843900445','8847500000'])]

excisetaxnbr major_minor  saleprice    recordingnbr volume page  \
41899        2975737  8843900445     815000  20190228001472               
149957       2975141  8843900445     699900  20190226001189               
162622       2993692  8847500000     764990  20190614000861               
197942       2976965  8843900445     706900  20190311000874               
210949       3020478  8847500000     698654  20191114000503               
292126       2980974  8843900445     799900  20190405000497               
294166       2975973  8843900445     782900  20190301000951               
312455       2975726  8843900445     776900  20190228001436               
332992       2976878  8847500000     784684  20190311000330               

       platnbr plattype         platlot platblock  ... saleinstrument  \
41899                                              ...              3   
149957                                             ...              3   
162622                                             ...              3   
197942                                             ...              3   
210949                                             ...              3   
292126                                             ...              3   
294166                                             ...              3   
312455                                             ...              3   
332992                                             ...              3   

       afforestland  afcurrentuseland  afnonprofituse  afhistoricproperty  \
41899             N                 N               N                   N   
149957            N                 N               N                   N   
162622            N                 N               N                   N   
197942            N                 N               N                   N   
210949            N                 N               N                   N   
292126            N                 N               N                   N   
294166            N                 N               N                   N   
312455            N                 N               N                   N   
332992            N                 N               N                   N   

       salereason propertyclass salewarning monthsold  yrsold  
41899           1             8                   2.0  2019.0  
149957          1             8                   2.0  2019.0  
162622          1             3                   6.0  2019.0  
197942          1             8                   3.0  2019.0  
210949          1             3                  11.0  2019.0  
292126          1             8                   3.0  2019.0  
294166          1             8                   2.0  2019.0  
312455          1             8                   2.0  2019.0  
332992          1             3          43       3.0  2019.0  

[9 rows x 24 columns]

In [19]:
#Dropping the duplicates
properties.drop_duplicates(subset='major_minor', keep='last', inplace=True)

<ipython-input-19-6281fb0af3f5>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  properties.drop_duplicates(subset='major_minor', keep='last', inplace=True)


In [20]:
properties.head()

excisetaxnbr major_minor  saleprice    recordingnbr volume page platnbr  \
9        3027422  2130430120     560000  20191226000848                       
10       3002257  9406520630     435000  20190730001339                       
27       3033993  9264902060     370000  20200214001331                       
50       2937935  8731780100     185000  20180625000560                       
55       2926991  2211700075    1425000  20180427001349                       

   plattype         platlot platblock  ... saleinstrument afforestland  \
9                                      ...              3            N   
10                                     ...              3            N   
27                                     ...              2            N   
50                                     ...              3            N   
55                                     ...              3            N   

    afcurrentuseland  afnonprofituse  afhistoricproperty salereason  \
9                  N               N                   N          1   
10                 N               N                   N          1   
27                 N               N                   N          1   
50                 N               N                   N          1   
55                 N               N                   N          1   

   propertyclass salewarning monthsold  yrsold  
9              8                  12.0  2019.0  
10             8                   7.0  2019.0  
27             8                   2.0  2020.0  
50             3                   6.0  2018.0  
55             8                   4.0  2018.0  

[5 rows x 24 columns]

In [21]:
properties.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45203 entries, 9 to 351062
Data columns (total 24 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   excisetaxnbr        45203 non-null  int64  
 1   major_minor         45203 non-null  object 
 2   saleprice           45203 non-null  int64  
 3   recordingnbr        45203 non-null  object 
 4   volume              45203 non-null  object 
 5   page                45203 non-null  object 
 6   platnbr             45203 non-null  object 
 7   plattype            45203 non-null  object 
 8   platlot             45203 non-null  object 
 9   platblock           45203 non-null  object 
 10  sellername          45203 non-null  object 
 11  buyername           45203 non-null  object 
 12  propertytype        45203 non-null  int64  
 13  principaluse        45203 non-null  int64  
 14  saleinstrument      45203 non-null  int64  
 15  afforestland        45203 non-null  object 
 16  afc

In [22]:
to_drop = ['volume', 'page', 'platnbr', 'plattype', 'platlot', 'platblock', 'sellername', 'buyername']
properties.drop(to_drop, axis=1, inplace=True)

C:\Users\15164\anaconda3\envs\learn-env\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [23]:
properties.head()

excisetaxnbr major_minor  saleprice    recordingnbr  propertytype  \
9        3027422  2130430120     560000  20191226000848            11   
10       3002257  9406520630     435000  20190730001339            11   
27       3033993  9264902060     370000  20200214001331            11   
50       2937935  8731780100     185000  20180625000560            14   
55       2926991  2211700075    1425000  20180427001349            11   

    principaluse  saleinstrument afforestland afcurrentuseland afnonprofituse  \
9              6               3            N                N              N   
10             6               3            N                N              N   
27             6               2            N                N              N   
50             2               3            N                N              N   
55             6               3            N                N              N   

   afhistoricproperty  salereason  propertyclass salewarning  monthsold  \
9                   N           1              8                   12.0   
10                  N           1              8                    7.0   
27                  N           1              8                    2.0   
50                  N           1              3                    6.0   
55                  N           1              8                    4.0   

    yrsold  
9   2019.0  
10  2019.0  
27  2020.0  
50  2018.0  
55  2018.0

Now I'll repeat everything I did for "properties" and apply it to residential buildings.

In [24]:
ResBldg.rename(columns=str.lower, inplace=True)
ResBldg['major'] = ResBldg['major'].astype(str)
ResBldg['minor'] = ResBldg['minor'].astype(str)

ResBldg['major'] = ResBldg['major'].apply(major_padded)
ResBldg['minor'] = ResBldg['minor'].apply(minor_padded)

ResBldg['major_minor'] = ResBldg['major']+ResBldg['minor']

col_name = 'major_minor'
last_col = ResBldg.pop(col_name)
ResBldg.insert(0, col_name, last_col)

ResBldg.head(2)

major_minor   major minor  bldgnbr  nbrlivingunits  \
0  0098000720  009800  0720        1               1   
1  0098020140  009802  0140        1               1   

                      address buildingnumber fraction directionprefix  \
0  27719  SE 26TH WAY   98075          27719                       SE   
1  2829   277TH TER SE  98075          2829                             

                  streetname  ... fpmultistory fpfreestanding fpadditional  \
0  26TH                       ...            0              0            0   
1  277TH                      ...            0              0            0   

   yrbuilt  yrrenovated  pcntcomplete  obsolescence  pcntnetcondition  \
0     2001            0             0             0                 0   
1     2004            0             0             0                 0   

   condition  addnlcost  
0          3          0  
1          3          0  

[2 rows x 51 columns]

In [25]:
sum(ResBldg.minor == '0000')

0

Since residential buildings don't have condominiums (**Minor=0000**), I will be forced to drop more data.

In [26]:
ResBldg.loc[1]

major_minor                           0098020140
major                                     009802
minor                                       0140
bldgnbr                                        1
nbrlivingunits                                 1
address               2829   277TH TER SE  98075
buildingnumber                             2829 
fraction                                        
directionprefix                                 
streetname             277TH                    
streettype                                TER   
directionsuffix                               SE
zipcode                                    98075
stories                                        2
bldggrade                                     10
bldggradevar                                   0
sqft1stfloor                                1610
sqfthalffloor                                  0
sqft2ndfloor                                1400
sqftupperfloor                                 0
sqftunfinfull       

In [27]:
ResBldg = ResBldg.dropna(how='any',axis=0)
ResBldg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 154594 entries, 0 to 181506
Data columns (total 51 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   major_minor         154594 non-null  object 
 1   major               154594 non-null  object 
 2   minor               154594 non-null  object 
 3   bldgnbr             154594 non-null  int64  
 4   nbrlivingunits      154594 non-null  int64  
 5   address             154594 non-null  object 
 6   buildingnumber      154594 non-null  object 
 7   fraction            154594 non-null  object 
 8   directionprefix     154594 non-null  object 
 9   streetname          154594 non-null  object 
 10  streettype          154594 non-null  object 
 11  directionsuffix     154594 non-null  object 
 12  zipcode             154594 non-null  object 
 13  stories             154594 non-null  float64
 14  bldggrade           154594 non-null  int64  
 15  bldggradevar        154594 non-nul

### Converting 'zipcode' to int

In [28]:
ResBldg['zipcode'] = ResBldg['zipcode'].str.split('-').str[0]
ResBldg['zipcode'].value_counts()

98042    4661
98023    4177
98115    4175
98038    3828
98058    3733
         ... 
98          1
90108       1
89045       1
98101       1
98085       1
Name: zipcode, Length: 122, dtype: int64

In [29]:
ResBldg['zipcode'] = ResBldg['zipcode'].astype(str).str.replace(',', '')
ResBldg['zipcode'] = ResBldg['zipcode'].astype(str).str.replace(' ', '')
ResBldg['zipcode'] = ResBldg['zipcode'].astype(str).str.replace('.', '')

In [30]:
df = ResBldg[ResBldg['zipcode'].apply(lambda x: not x.isnumeric())]
df

major_minor   major minor  bldgnbr  nbrlivingunits  \
71823   3126069112  312606  9112        1               1   
173514  9828702305  982870  2305        1               1   

                              address buildingnumber fraction directionprefix  \
71823        11026   AVONDALE RD NE            11026                            
173514  111 D  26TH AVE E  A on plans          111        D                     

                       streetname  ... fpmultistory fpfreestanding  \
71823   AVONDALE                   ...            0              0   
173514  26TH                       ...            0              0   

       fpadditional  yrbuilt  yrrenovated  pcntcomplete  obsolescence  \
71823             0     2015            0             0             0   
173514            0     2019            0            60             0   

        pcntnetcondition  condition  addnlcost  
71823                  0          3       8000  
173514                 0          3          0  

[2 rows x 51 columns]

In [31]:
#Dropping non-numeric zipcode entries
ResBldg = ResBldg.drop(index=[71823,173514])

In [32]:
ResBldg['zipcode'].value_counts()

98042    4661
98023    4177
98115    4175
98038    3828
98058    3756
         ... 
98085       1
89045       1
90333       1
98083       1
98026       1
Name: zipcode, Length: 119, dtype: int64

In [33]:
ResBldg['zipcode']= ResBldg['zipcode'].astype(np.int64)

In [34]:
prop_res = properties.merge(ResBldg, on='major_minor', suffixes=("", " "))
prop_res.describe()

excisetaxnbr     saleprice  propertytype  principaluse  saleinstrument  \
count  3.392800e+04  3.392800e+04  33928.000000       33928.0    33928.000000   
mean   2.987995e+06  7.676240e+05     11.024228           6.0        3.073450   
std    3.762621e+04  5.874356e+05      0.280116           0.0        2.185849   
min    2.909403e+06  5.000000e+00     11.000000           6.0        2.000000   
25%    2.956618e+06  4.500000e+05     11.000000           6.0        3.000000   
50%    2.992006e+06  6.450000e+05     11.000000           6.0        3.000000   
75%    3.019295e+06  8.850000e+05     11.000000           6.0        3.000000   
max    3.050093e+06  2.675000e+07     18.000000           6.0       27.000000   

         salereason  propertyclass     monthsold        yrsold       bldgnbr  \
count  33928.000000        33928.0  33928.000000  33928.000000  33928.000000   
mean       1.014914            8.0      6.396781   2018.826839      1.008872   
std        0.501786            0.0      3.174764      0.708859      0.113924   
min        1.000000            8.0      1.000000   2018.000000      1.000000   
25%        1.000000            8.0      4.000000   2018.000000      1.000000   
50%        1.000000            8.0      6.000000   2019.000000      1.000000   
75%        1.000000            8.0      9.000000   2019.000000      1.000000   
max       18.000000            8.0     12.000000   2020.000000      7.000000   

       ...  fpmultistory  fpfreestanding  fpadditional       yrbuilt  \
count  ...  33928.000000    33928.000000  33928.000000  33928.000000   
mean   ...      0.337921        0.078961      0.198744   1968.021752   
std    ...      0.512201        0.280399      0.418317     28.385749   
min    ...      0.000000        0.000000      0.000000   1900.000000   
25%    ...      0.000000        0.000000      0.000000   1950.000000   
50%    ...      0.000000        0.000000      0.000000   1969.000000   
75%    ...      1.000000        0.000000      0.000000   1990.000000   
max    ...      5.000000        3.000000      3.000000   2020.000000   

        yrrenovated  pcntcomplete  obsolescence  pcntnetcondition  \
count  33928.000000  33928.000000  33928.000000      33928.000000   
mean     110.160133      0.173898      0.110410          0.006425   
std      456.266456      3.498123      2.103985          0.457740   
min        0.000000      0.000000      0.000000          0.000000   
25%        0.000000      0.000000      0.000000          0.000000   
50%        0.000000      0.000000      0.000000          0.000000   
75%        0.000000      0.000000      0.000000          0.000000   
max     2020.000000    100.000000     99.000000         43.000000   

          condition     addnlcost  
count  33928.000000  33928.000000  
mean       3.536401    597.927995  
std        0.718633   2286.108745  
min        1.000000      0.000000  
25%        3.000000      0.000000  
50%        3.000000      0.000000  
75%        4.000000      0.000000  
max        5.000000  50000.000000  

[8 rows x 48 columns]

### Converting the 'yrrenovated' column to a binary
Next, the 'yrrenovated' column has a large amount zero values (denoting that a house hasn't been renovated). Rather than dropping the data, I will make this a binary column whether a house was sold within 10 years of a renovation or not, or built within the last five years. 

First, I will make a new column 'age' with the properties' ages as variables to get a better sense of how old a home is.

In [35]:
#Convert YrBuilt to age by subtracting the build year from the most recent year in the in dataframe
prop_res['age'] = 2020 - prop_res.yrbuilt

Now I will convert the 'yrrenovated' column to a binary column.

In [36]:
prop_res['yrrenovated'].value_counts()

0       32059
2018      173
2019      122
2017       66
2003       58
        ...  
1935        1
1949        1
1944        1
1952        1
1936        1
Name: yrrenovated, Length: 80, dtype: int64

In [37]:
prop_res['yrrenovated'].fillna(0.0, inplace=True)

In [38]:
prop_res['renovated'] = prop_res['yrsold'] - prop_res['yrrenovated']

In [39]:
#Replacing values less than 10 with 1, and values over 10 with 0
Renovated = prop_res.renovated.values
Age = prop_res.age.values
values = np.where(Renovated <= 10, 1, 0)
prop_res['renovated'] = np.where(Age <= 5, 1, values)

Now I'll check again for duplicates:

In [40]:
prop_res.major_minor.value_counts()

0809003030    5
6373000170    5
7129303601    4
2822059059    4
6196600360    4
             ..
8651402910    1
3276980010    1
1822059374    1
3432501415    1
3352400905    1
Name: major_minor, Length: 33671, dtype: int64

In [41]:
prop_res[prop_res['major_minor'].isin(['0809003030', '6373000170', '2822059059'])]

excisetaxnbr major_minor  saleprice    recordingnbr  propertytype  \
1886        3048606  0809003030    1176000  20200527001133            11   
1887        3048606  0809003030    1176000  20200527001133            11   
1888        3048606  0809003030    1176000  20200527001133            11   
1889        3048606  0809003030    1176000  20200527001133            11   
1890        3048606  0809003030    1176000  20200527001133            11   
24272       2921369  6373000170    1500000  20180328000779            11   
24273       2921369  6373000170    1500000  20180328000779            11   
24274       2921369  6373000170    1500000  20180328000779            11   
24275       2921369  6373000170    1500000  20180328000779            11   
24276       2921369  6373000170    1500000  20180328000779            11   
24803       2933171  2822059059     825000  20180531000805            11   
24804       2933171  2822059059     825000  20180531000805            11   
24805       2933171  2822059059     825000  20180531000805            11   
24806       2933171  2822059059     825000  20180531000805            11   

       principaluse  saleinstrument afforestland afcurrentuseland  \
1886              6               3            N                N   
1887              6               3            N                N   
1888              6               3            N                N   
1889              6               3            N                N   
1890              6               3            N                N   
24272             6               3            N                N   
24273             6               3            N                N   
24274             6               3            N                N   
24275             6               3            N                N   
24276             6               3            N                N   
24803             6               3            N                N   
24804             6               3            N                N   
24805             6               3            N                N   
24806             6               3            N                N   

      afnonprofituse  ... fpadditional  yrbuilt  yrrenovated pcntcomplete  \
1886               N  ...            0     2019            0           40   
1887               N  ...            0     2019            0           40   
1888               N  ...            0     2019            0           40   
1889               N  ...            0     2019            0           40   
1890               N  ...            0     2019            0           40   
24272              N  ...            0     2019            0           60   
24273              N  ...            0     2019            0           60   
24274              N  ...            0     2019            0           60   
24275              N  ...            0     2019            0           60   
24276              N  ...            0     2019            0           60   
24803              N  ...            1     1949            0            0   
24804              N  ...            0     1924            0            0   
24805              N  ...            0     1934         2010            0   
24806              N  ...            0     1981            0            0   

       obsolescence  pcntnetcondition condition addnlcost  age  renovated  
1886              0                 0         3         0    1          1  
1887              0                 0         3         0    1          1  
1888              0                 0         3         0    1          1  
1889              0                 0         3         0    1          1  
1890              0                 0         3         0    1          1  
24272             0                 0         3         0    1          1  
24273             0                 0         3         0    1          1  
24274             0                 0         3         0    1    

In [42]:
prop_res.drop_duplicates(subset='major_minor', inplace=True)
len(prop_res)

33671

We are now left with 33,671 entries in our data frame.

### Now for our last file, 'Parcel':

In [43]:
Parcel.rename(columns=str.lower, inplace=True)
Parcel['major'] = Parcel['major'].astype(str)
Parcel['minor'] = Parcel['minor'].astype(str)
Parcel['major'] = Parcel['major'].apply(major_padded)
Parcel['minor'] = Parcel['minor'].apply(minor_padded)
Parcel['major_minor'] = Parcel['major']+Parcel['minor']
last_col = Parcel.pop(col_name)
Parcel.insert(0, col_name, last_col)

In [44]:
prop_res = prop_res.merge(Parcel, on='major_minor', suffixes=("", " "))

In [45]:
prop_res.shape

(33671, 150)

In [46]:
prop_res['proptype'].unique()

array(['R'], dtype=object)

Thankfully, we haven't lost any more entries *and* our property type column only contains 'R' (residential) values.

In [47]:
prop_res.describe()

excisetaxnbr     saleprice  propertytype  principaluse  saleinstrument  \
count  3.367100e+04  3.367100e+04  33671.000000       33671.0     33671.00000   
mean   2.987987e+06  7.654065e+05     11.023908           6.0         3.07110   
std    3.760185e+04  5.794149e+05      0.280319           0.0         2.17568   
min    2.909403e+06  5.000000e+00     11.000000           6.0         2.00000   
25%    2.956624e+06  4.500000e+05     11.000000           6.0         3.00000   
50%    2.992018e+06  6.450000e+05     11.000000           6.0         3.00000   
75%    3.018590e+06  8.840000e+05     11.000000           6.0         3.00000   
max    3.050093e+06  2.675000e+07     18.000000           6.0        27.00000   

         salereason  propertyclass     monthsold        yrsold       bldgnbr  \
count  33671.000000        33671.0  33671.000000  33671.000000  33671.000000   
mean       1.015028            8.0      6.402958   2018.826141      1.003772   
std        0.503696            0.0      3.172576      0.708517      0.071995   
min        1.000000            8.0      1.000000   2018.000000      1.000000   
25%        1.000000            8.0      4.000000   2018.000000      1.000000   
50%        1.000000            8.0      6.000000   2019.000000      1.000000   
75%        1.000000            8.0      9.000000   2019.000000      1.000000   
max       18.000000            8.0     12.000000   2020.000000      7.000000   

       ...  wfntpoorquality  wfntrestrictedaccess  tidelandshoreland  \
count  ...     33671.000000          33671.000000       33671.000000   
mean   ...         0.000713              0.005554           0.015117   
std    ...         0.026689              0.108990           0.172277   
min    ...         0.000000              0.000000           0.000000   
25%    ...         0.000000              0.000000           0.000000   
50%    ...         0.000000              0.000000           0.000000   
75%    ...         0.000000              0.000000           0.000000   
max    ...         1.000000              3.000000           3.000000   

       lotdepthfactor  trafficnoise  airportnoise  nbrbldgsites  \
count    33671.000000  33671.000000       33671.0   33671.00000   
mean         0.049004      0.207270           0.0       0.00689   
std          2.031641      0.554123           0.0       0.14849   
min          0.000000      0.000000           0.0       0.00000   
25%          0.000000      0.000000           0.0       0.00000   
50%          0.000000      0.000000           0.0       0.00000   
75%          0.000000      0.000000           0.0       0.00000   
max        109.000000      3.000000           0.0      16.00000   

       contamination  historicsite  currentusedesignation  
count        33671.0  33671.000000           33671.000000  
mean             0.0      0.001277               0.002762  
std              0.0      0.060917               0.087662  
min              0.0      0.000000               0.000000  
25%              0.0      0.000000               0.000000  
50%              0.0      0.000000               0.000000  
75%              0.0      0.000000               0.000000  
max              0.0      3.000000               3.000000  

[8 rows x 94 columns]

**I will select features of a residential property that would have a bearing on its saleprice:**

In [48]:
property_data = prop_res.loc[:, ['saleprice', 'sqfttotliving', 'yrbuilt', 'yrrenovated', 'bedrooms',
                      'zipcode', 'sqfttotbasement', 'sqftfinbasement', 'sqftopenporch',
                      'sqftenclosedporch', 'sqftdeck','heatsystem', 'heatsource', 'bathhalfcount', 
                      'bath3qtrcount', 'bathfullcount','condition', 'viewutilization', 'sqftgarageattached',
                      'daylightbasement', 'finbasementgrade', 'hbuasifvacant', 'inadequateparking', 'township',
                      'mtrainier', 'olympics', 'cascades', 'territorial', 'seattleskyline', 'pugetsound',
                      'lakewashington', 'lakesammamish', 'smalllakerivercreek', 'otherview', 'wfntlocation',
                      'trafficnoise', 'airportnoise', 'powerlines', 'othernuisances', 'adjacentgreenbelt']]

In [49]:
property_data.head()

saleprice  sqfttotliving  yrbuilt  yrrenovated  bedrooms  zipcode  \
0     560000           1930     1989            0         3    98019   
1     435000           1975     1994            0         3    98038   
2     370000           1750     1979            0         3    98023   
3    1425000           2850     1985            0         4    98006   
4    1085000           1870     2008            0         3    98102   

   sqfttotbasement  sqftfinbasement  sqftopenporch  sqftenclosedporch  ...  \
0                0                0            180                  0  ...   
1                0                0             45                  0  ...   
2             1130              460              0                  0  ...   
3                0                0            190                  0  ...   
4              790              790              0                  0  ...   

   lakewashington  lakesammamish  smalllakerivercreek  otherview  \
0               0              0                    0          0   
1               0              0                    0          0   
2               0              0                    0          0   
3               0              0                    0          0   
4               0              0                    0          0   

   wfntlocation  trafficnoise  airportnoise powerlines  othernuisances  \
0             0             0             0          N               N   
1             0             0             0          N               N   
2             0             0             0          N               N   
3             0             0             0          N               N   
4             0             0             0          N               N   

  adjacentgreenbelt  
0                 N  
1                 N  
2                 N  
3                 N  
4                 N  

[5 rows x 40 columns]

In [50]:
cleaned_data = property_data.to_csv('cleaned_data.csv')
property_data = prop_res.to_csv('property_data.csv')

In [1]:
kc_data = pd.read_csv('../../data/processed/cleaned_data.csv')
kc_data.head()

NameError: name 'pd' is not defined